ext/liquid_c/context.c

```diff
@@ -0,0 +1,129 @@
+#include "liquid.h"
+#include "context.h"
+#include "variable_lookup.h"
+
+static VALUE cLiquidVariableLookup, cLiquidUndefinedVariable;
+ID id_aset, id_call, id_to_liquid, id_set_context;
+static ID id_evaluate, id_has_key, id_aref;
+static ID id_ivar_scopes, id_ivar_environments, id_ivar_strict_variables;
+
+VALUE context_evaluate(VALUE self, VALUE expression)
+{
+    // Scalar type stored directly in the VALUE
+    if (RB_SPECIAL_CONST_P(expression))
```

Comment:
Yeah, this just has to look at the `VALUE` and turns into a nearly free check. The start of this function ends up looking a lot like [`rb_class_of`](https://github.com/ruby/ruby/blob/v2_6_2/include/ruby/ruby.h#L2056), because of `#define RB_SPECIAL_CONST_P(x) (RB_IMMEDIATE_P(x) || !RB_TEST(x))`.